### PR TITLE
Fix PHP 8.0 Issues

### LINKED ISSUE #3210: Perform PHP 8.0 Audit
Yikes, November 26th arrived much earlier this year than it has in the past!!!

PHP 8.0 with a slew of changes was released today so we need to begin preparing for customers to begin switching over. I don't actually think that will happen any time soon because this version of PHP has a number of backwards compatibility breaking changes and I don't expect most people to be prepared any time soon, at least not with production servers. That said, PHP 8 introduces JIT (just in time) compilation which has the potential to increase performance significantly (although I doubt much for WordPRess sites) so there may be some pushing to try it out. 

[WordPress Core aims to be compatible with PHP 8.0 in the 5.6 release (currently scheduled for December 8, 2020)](https://make.wordpress.org/core/2020/11/23/wordpress-and-php-8-0/) so we likely have at least a couple of weeks before we will even see any mentions of whether we are compatible.

### LINKED ISSUE #3521: PHP 8 Issues
Upgraded my localhost to PHP 8 in order to troubleshoot #3515 and discovered the following issues as well:

-
   ```
   Deprecated: Required parameter $pan_message follows optional parameter $pan_name
   in /core/EE_Error.core.php on line 1159
   ```

-
   ```
   Deprecated: Required parameter $capabilities_checker follows optional parameter $return_url
   in /core/services/notifications/PersistentAdminNoticeManager.php on line 68
   ```

-
   ```
   Deprecated: Required parameter $request follows optional parameter $return_url
   in /core/services/notifications/PersistentAdminNoticeManager.php on line 68
   ```

-
   ```
   Fatal error: Uncaught TypeError:
   EventEspresso\core\services\notifications\PersistentAdminNoticeManager::__construct():
   Argument #2 ($capabilities_checker) must be of type
   EventEspresso\core\domain\services\capabilities\CapabilitiesChecker
   null given in /core/services/notifications/PersistentAdminNoticeManager.php:68
   ```

-
   ```
   Deprecated: Required parameter $old_value follows optional parameter $option
   in /core/EE_Config.core.php on line 402
   ```

-
   ```
   Deprecated: Required parameter $value follows optional parameter $option 
   in /core/EE_Config.core.php on line 402
   ```

-
   ```
   Deprecated: Required parameter $registration follows optional parameter $EVT_ID 
   in /core/db_models/EEM_Event.model.php on line 449
   ```

-
   ```
   Deprecated: Required parameter $model_field_factory follows optional parameter $timezone
   in /core/db_models/EEM_Attendee.model.php on line 137
   ```

-
   ```
   Deprecated: Required parameter $model_field_factory follows optional parameter $timezone 
   in /core/db_models/EEM_WP_User.model.php on line 32
   ```

-
   ```
   Deprecated: Required parameter $reg_step follows optional parameter $state_options 
   in /modules/add_new_state/EED_Add_New_State.module.php on line 707
   ```

-
   ```
   Deprecated: Required parameter $registration follows optional parameter $state_options 
   in /modules/add_new_state/EED_Add_New_State.module.php on line 707
   ```

-
   ```
   Deprecated: Required parameter $question follows optional parameter $state_options 
   in /modules/add_new_state/EED_Add_New_State.module.php on line 707
   ```

-
   ```
   Deprecated: Required parameter $answer follows optional parameter $state_options 
   in /modules/add_new_state/EED_Add_New_State.module.php on line 707
   ```

-
   ```
   Deprecated: Required parameter $reg_step follows optional parameter $country_options 
   in /modules/add_new_state/EED_Add_New_State.module.php on line 749
   ```

-
   ```
   Deprecated: Required parameter $registration follows optional parameter $country_options 
   in /modules/add_new_state/EED_Add_New_State.module.php on line 749
   ```

-
   ```
   Deprecated: Required parameter $question follows optional parameter $country_options 
   in /modules/add_new_state/EED_Add_New_State.module.php on line 749
   ```

-
   ```
   Deprecated: Required parameter $answer follows optional parameter $country_options 
   in /modules/add_new_state/EED_Add_New_State.module.php on line 749
   ```

-
   ```
   Deprecated: Required parameter $transaction follows optional parameter $class 
   in /modules/messages/EED_Messages.module.php on line 1316
   ```

-
   ```
   Deprecated: Required parameter $ticket follows optional parameter $qty 
   in /modules/ticket_sales_monitor/EED_Ticket_Sales_Monitor.module.php on line 249
   ```

-
   ```
   Deprecated: Required parameter $source follows optional parameter $valid_reserved_ticket_line_items 
   in /modules/ticket_sales_monitor/EED_Ticket_Sales_Monitor.module.php on line 992
   ```

-
   ```
   Deprecated: Required parameter $model_name follows optional parameter $default_value 
   in /core/services/orm/ModelFieldFactory.php on line 95
   ```

-
   ```
   Deprecated: Required parameter $allowed_enum_values follows optional parameter $default_value 
   in /core/services/orm/ModelFieldFactory.php on line 226
   ```

-
   ```
   Deprecated: Required parameter $domain follows optional parameter $i18n_map 
   in /core/services/assets/I18nRegistry.php on line 63
   ```

-
   ```
   Deprecated: Required parameter $ticket follows optional parameter $prices 
   in /caffeinated/admin/new/pricing/espresso_events_Pricing_Hooks.class.php on line 873
   ```

-
   ```
   Deprecated: Required parameter $default follows optional parameter $datetime_tickets 
   in /caffeinated/admin/new/pricing/espresso_events_Pricing_Hooks.class.php on line 1279
   ```

-
   ```
   Deprecated: Required parameter $default follows optional parameter $datetime_tickets 
   in /caffeinated/admin/new/pricing/espresso_events_Pricing_Hooks.class.php on line 1347
   ```

-
   ```
   Deprecated: Required parameter $default follows optional parameter $ticket_datetimes 
   in /caffeinated/admin/new/pricing/espresso_events_Pricing_Hooks.class.php on line 2026
   ```

-
   ```
   Fatal error: Uncaught Error: Unknown named parameter $new_price
    in /core/admin/EE_Admin_Page.core.php:1010 
   	```

-
   ```
   Deprecated: Required parameter $registry follows optional parameter $form_action
   in /core/libraries/form_sections/form_handlers/FormHandler.php on line 134
   ```

-
   ```
   Deprecated:  Required parameter $file follows optional parameter $dismissible
   in /core/services/notices/NoticesContainer.php on line 92
   ```

-
   ```
   Deprecated: Required parameter $fieldname follows optional parameter $what 
   in /core/db_classes/EE_Base_Class.class.php on line 1572
   ```

### PR DESCRIPTION
**Pull request by tn3rb**
plz see: #3515 && #3521 & #3851

This PR:

- fixes some more named parameter mismatches caused by using `call_user_func_array()`

- fixes a bunch of required parameters positioned after optional parameters, sometimes by also making them optional (if not necessary to function or appropriate checks in place) or by making the optional parameters before them required parameters (if appropriate checks already in place)

- ensures that some classes are setup properly prior to loading, now that optional parameters are required

- changes how some dependency resolution occurs in `EE_Registry` due to changes in how PHP 8's Reflection classes report information about code.

- closes #3210  closes #3521

### CODE REVIEW

**Comment by tn3rb** [comment]
will fix tests tomorrow

**Comment by tn3rb** [comment]
Compatibility fixes for PHP 8.1 will have to happen in another branch as they are not compatible with earlier versions

so we either have to:

- wait until WP core updates to PHP 8.1 :nauseated_face:

- create a long living duplicate branch that is updated to be 8.1 compatible :nauseated_face: :nauseated_face:

File: tests/testcases/core/admin/EE_Admin_Tests.php [hunk 1]
```diff
@@ -222,7 +222,7 @@ public function test_dashboard_glance_items() {
 
 		$generated_items = EE_Admin::instance()->dashboard_glance_items( array() );
 		//first assert the elements are an array.
-		$this->assertInternalType('array', $generated_items);
+		$this->assertIsArray( $generated_items);
```
Comment (hoseinrafiei) [inline tests/testcases/core/admin/EE_Admin_Tests.php:225]: Found an extra space here 😂

File: tests/testcases/core/db_models/EEM_Datetime_Test.php [hunk 2]
```diff
@@ -453,7 +453,7 @@ public function test_get_dtt_months_and_years() {
 	public function test_get_datetime_counts_by_status_and_get_datetime_count_for_status() {
 
 	    $existing_datetimes = EEM_Datetime::instance()->get_all();
-        $this->assertInternalType('array', $existing_datetimes);
+        $this->assertIsArray( $existing_datetimes);
```
Comment (hoseinrafiei) [inline tests/testcases/core/db_models/EEM_Datetime_Test.php:456]: Another one here

**Review by hoseinrafiei — APPROVED** [review]
Finally finished this huge PR. It was perfect 💯

File: core/EE_Error.core.php [hunk 3]
```diff
@@ -665,7 +649,7 @@ private static function _add_notice($type = 'success', $msg = '', $file = '', $f
      */
     public static function overwrite_success()
     {
-        self::$_espresso_notices['success'] = false;
+        self::$_espresso_notices['success'] = [];
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply core/EE_Error.core.php:652]: switching data types === BAD

File: core/db_classes/EE_Attendee.class.php [hunk 4]
```diff
@@ -9,8 +9,8 @@
  * @ package            Event Espresso
  * @ author                Seth Shoultes
  * @ copyright        (c) 2008-2011 Event Espresso  All Rights Reserved.
- * @ license            {@link https://eventespresso.com/support/terms-conditions/}   * see Plugin Licensing *
- * @ link                    {@link http://www.eventespresso.com}
+ * @linkense            {@link https://eventespresso.com/support/terms-conditions/}   * see Plugin Licensing *
+ * @linkk                    {@link http://www.eventespresso.com}
```
[the remark this answers is not in the document]
Comment (tn3rb) [reply core/db_classes/EE_Attendee.class.php:13]: lolz...  thnx will fix

File: tests/testcases/core/admin/EE_Admin_Tests.php (hunk 1 above)
Comment (tn3rb) [reply tests/testcases/core/admin/EE_Admin_Tests.php:225]: thnx will fix

File: tests/testcases/core/db_models/EEM_Datetime_Test.php (hunk 2 above)
Comment (tn3rb) [reply tests/testcases/core/db_models/EEM_Datetime_Test.php:456]: thnx will fix

File: core/EE_Error.core.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (hoseinrafiei) [reply core/EE_Error.core.php:597]: Agree

**Pull request merged**